machine/machine_core/ssh_connection.py

```diff
@@ -248,26 +255,28 @@ def _check_ssh_master(self):
                 return True
         return False
 
-    def _ensure_ssh_master(self):
+    def _ensure_ssh_master(self) -> None:
         if not self._check_ssh_master():
             self._start_ssh_master()
 
-    def __ssh_direct_opt_var(self, direct=False):
+    def __ssh_direct_opt_var(self, direct: bool = False) -> bool:
         return bool(os.getenv("TEST_SSH_DIRECT", direct))
 
-    def __execution_opts(self, direct=False):
+    def __execution_opts(self, direct: bool = False) -> List[str]:
         direct = self.__ssh_direct_opt_var(direct=direct)
         if direct:
             return ["-i", self.identity_file]
-        else:
+        elif self.ssh_master:
             return ["-o", "ControlPath=" + self.ssh_master]
+        return []
```

Comment:
Tricky.  If you're going to take this approach to avoid the warning about `self.ssh_master` possibly being `None` then I'd prefer to see some kind of an error raised here instead of returning empty.